sdk/core/azure-core/inc/azure/core/internal/io/file_helpers.hpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/**
+ * @file
+ * @brief Internal utility functions for files and directories.
+ *
+ */
+#pragma once
+
+#include <azure/core/platform.hpp>
+
+#include <stdexcept>
+#include <string>
+
+#include <sys/stat.h> // for stat() used to check file size and mkdir() used to create directories
+
+#if defined(AZ_PLATFORM_WINDOWS)
+#if !defined(WIN32_LEAN_AND_MEAN)
+#define WIN32_LEAN_AND_MEAN
+#endif
+#if !defined(NOMINMAX)
```

Comment:
As I mentioned, we don't ever want to #define NOMINMAX because it alters the preprocessor definitions created by windows.h.

And in general, I think we should avoid platform specific headers if at all possible - it's better to define an API which is platform agnostic and then put the platform specific implementation in a .cpp file.

That way including our headers doesn't pollute the customer's global namespace.

Reply:
I respectfully but passionately disagree - https://github.com/Azure/azure-sdk-for-cpp/issues/5660#issuecomment-2142854964.

Reply:
But I fully support your idea about moving the implementation to `.cpp`, and then we don't have to include Windows.h in this header, and there is no need for `NOMINMAX` in the header that we need to discuss.

I'd still recommend to `#define NOMINMAX` and `WIN32_LEAN_AND_MEAN` before including windows.h in the `file_helpers.cpp`.

Reply:
In a source file, I honestly don't care that much.

But in a header, I care passionately. 

I've been thinking about this and I'm going to write a guideline for it in my next round of guidelines. 

Specifically I'm considering adding a rule:

- Azure headers MUST NOT define types outside of the Azure namespace.

What this effectively does is to prevent any Azure headers from exposing any types defined outside Azure, and in practice that blocks including headers like libcurl.h, openssl/*, windows.h, etc, since those all define types in the global namespace.

Reply:
I'll be moving the implementation into the source along with the windows.h header, to help resolve the feedback.

The conservation is good. Given this impacts the SDK as a whole and guidelines, please move it over to the issue, so the details aren't lost:
https://github.com/Azure/azure-sdk-for-cpp/issues/5660